app/services/course/user_invitation_service.rb

```diff
@@ -63,13 +65,25 @@ def resend_invitation(invitations)
   #     Array<(Array<Course::UserInvitation>,
   #     Array<Course::UserInvitation>,
   #     Array<CourseUser>,
-  #     Array<CourseUser>)>
+  #     Array<CourseUser>)>,
+  #     Array<Hash>,
   #   ]
   #   A tuple containing the users newly invited, already invited,
-  #     newly registered and already registered respectively.
+  #     newly registered and already registered, and duplicate users respectively.
   # @raise [CSV::MalformedCSVError] When the file provided is invalid.
   def invite_users(users)
-    users = parse_invitations(users)
-    process_invitations(users)
+    users, duplicate_users = parse_invitations(users)
```

Comment:
Instead of manipulating the non-unique users here, how about partitioning differently in `#parse_invitations`:

```ruby
users.partition.with_index do |user, index|
  users.find_index( { |u| u.email == user.email } == index
end
```

Alternative:
```ruby
unique_users = users.uniq(&:email)
unique_users, users - unique_users
```

Reply:
`users - unique_users` actually results in an empty array as all occurrences of each entry in `unique_users` is removed.